jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,665 @@
+     */
+    public void excludePath(String... paths)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        _paths.exclude(paths);
+    }
+
+    /**
+     * Include {@link InetAddressPattern}s in the conditions to be met
+     * @param patterns {@link InetAddressPattern}s that are
+     *              tested against the {@link ConnectionMetaData#getRemoteSocketAddress() getRemoteSocketAddress()} of
+     *              {@link Request#getConnectionMetaData()}.
+     */
+    public void include(InetAddressPattern... patterns)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (InetAddressPattern p : patterns)
+            _predicates.include(new InetAddressPatternPredicate(p));
+    }
+
+    /**
+     * Include {@link InetAddressPattern}s in the conditions to be met
+     * @param patterns String representations of {@link InetAddressPattern}s that are
+     *              tested against the {@link ConnectionMetaData#getRemoteSocketAddress() getRemoteSocketAddress()} of
+     *              {@link Request#getConnectionMetaData()}.
+     */
+    public void includeInetAddressPattern(String... patterns)
+    {
+        for (String p : patterns)
+            include(InetAddressPattern.from(p));
+    }
+
+    /**
+     * Exclude {@link InetAddressPattern}s in the conditions to be met
+     * @param patterns {@link InetAddressPattern}s that are
+     *              tested against the {@link ConnectionMetaData#getRemoteSocketAddress() getRemoteSocketAddress()} of
+     *              {@link Request#getConnectionMetaData()}.
+     */
+    public void exclude(InetAddressPattern... patterns)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (InetAddressPattern p : patterns)
+            _predicates.exclude(new InetAddressPatternPredicate(p));
+    }
+
+    /**
+     * Exclude {@link InetAddressPattern} in the conditions to be met
+     * @param patterns String representations of {@link InetAddressPattern}s that are
+     *              tested against the {@link ConnectionMetaData#getRemoteSocketAddress() getRemoteSocketAddress()} of
+     *              {@link Request#getConnectionMetaData()}.
+     */
+    public void excludeInetAddressPattern(String... patterns)
+    {
+        for (String p : patterns)
+            exclude(InetAddressPattern.from(p));
+    }
+
+    /**
+     * {@link IncludeExclude#include(Object) Include} arbitrary {@link Predicate}s in the conditions.
+     * @param predicates {@link Predicate}s that are tested against the {@link Request}.
+     * This method is optimized so that a passed {@link MethodPredicate} or {@link PathSpecPredicate} is
+     * converted to a more efficient {@link #includeMethod(String...)} or {@link #include(PathSpec...)} respectively.
+     */
+    @SafeVarargs
+    public final void include(Predicate<Request>... predicates)
```

Comment:
The idea of an IncludeExcludeSet of `Predicate<Request>` is bizarre.
(and totally against the grain when it comes to normal Predicate usage)

This entire _predicates field should just be dropped in favor of a simple `Predicate<Request>`  let the users build up anything in the way they want.
Without having to mentally juggle test + include + exclude at the same time.

If this exists for InetAddress behaviors, then make _that_ be an public API (eg: `InetAddressPredicate`) for building a single `Predicate<Request>` based on `IncludeExcludeSet` of addresses.  Then the user can build up that Predicate with any other combinations they want (with and/or/not logic of their choice).

This `IncludeExcludeSet<Predicate<Request>, Request>` just eliminates the normal ability to chain Predicates together with and/or/not logic.

Reply:
@joakime it is not bizarre.  The whole point of the concept of "Predicate" is to decompose logic into understandable chunks, which can then be combined with operations like and/or/not etc.     The `IncludeExclude` class is just a utility method that combines `Predicate` collections in a known and standardised way.   It is itself a `Predicate` and (as you have helpfully pointed out elsewhere in your review) can be easily combined with other predicates with and/or/not logic.  It's javadoc describes it combination in terms of standard and/or/not operations as "the passed object is not in the excluded set AND ( either the included set is empty OR the object is in the included set)", i.e. it is
```
    !exclude0.test(request) && ... && !excludeN.test(request) && ( M == 0 || include0.test(request) || ... || includeM.test(request))
```
whilst not hugely complex, it is easy enough to make a mistake in such logic, as we did many times before the introduction of `IncludeExcludeSet`.

If users do not want to use that construct, then they can simply add their own complex predicate as a single include predicate and the class devolves to just `include0.test(request)`.

Reply:
Note also, that by using specific type of `IncludeExclude` (such as for String or PathSpecs), then we are able to replace linear logic like `exclude0.test(request) && ... && !excludeN.test(request)` with a single `excludeSet.contains(request)`, which can have an O(1) lookup.

Reply:
@joakime I had a look to see if I could use `Predicate.or(Predicate)` as a way of internally combining sets of `Predicates`.   It does work, but would result in needlessly deep stacks if we have many predicates.